### PR TITLE
add 32bit version

### LINKED ISSUE #33: [feature request] 32-bit compilation
Would you be willing to support 32bit compilation? For ARM targets, this would be important.

FWIW, while trying to compile 32 bit I get this with vs2019:

```shellsession
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(68,7): error C3861: '_BitScanReverse64': identifier not found [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(84,66): error C2169: 'fast_float::__emulu': intrinsic function, cannot be defined [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(89,17): error C2065: '__emulu': undeclared identifier [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(89): error C7552: 'fast_float::__emulu': purely intrinsic functions have no address [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(89,60): error C2440: 'initializing': cannot convert from 'unsigned __int64 (__cdecl *)(unsigned int,unsigned int)' to 'uint64_t' [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(89,15): message : There is no context in which this conversion is possible [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(90,17): error C2065: '__emulu': undeclared identifier [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(90): error C7552: 'fast_float::__emulu': purely intrinsic functions have no address [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(90,52): error C2440: 'initializing': cannot convert from 'unsigned __int64 (__cdecl *)(unsigned int,unsigned int)' to 'uint64_t' [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(90,15): message : There is no context in which this conversion is possible [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(91,24): error C2065: '__emulu': undeclared identifier [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(91,67): error C2568: '+': unable to resolve function overload [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(91,67): message : could be 'unsigned __int64 fast_float::__emulu(unsigned int,unsigned int)' [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(94,9): error C2065: '__emulu': undeclared identifier [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
D:\a\c4core\c4core\src\c4\ext\fast_float\include\fast_float\float_common.h(94,76): error C2296: '+': illegal, left operand has type 'unsigned __int64 (__cdecl *)(unsigned int,unsigned int)' [D:\a\c4core\c4core\build\static32\c4core.vcxproj]
```
and on gcc, I get this:

```shellsession
[  3%] Building CXX object CMakeFiles/c4core.dir/src/c4/base64.cpp.o
[  3%] Building CXX object CMakeFiles/c4core.dir/src/c4/char_traits.cpp.o
[  5%] Building CXX object CMakeFiles/c4core.dir/src/c4/error.cpp.o
In file included from /home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/ascii_number.h:9:0,
                 from /home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/parse_number.h:3,
                 from /home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/fast_float.h:44,
                 from /home/runner/work/c4core/c4core/src/c4/charconv.hpp:21,
                 from /home/runner/work/c4core/c4core/src/c4/base64.hpp:9,
                 from /home/runner/work/c4core/c4core/src/c4/base64.cpp:1:
/home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/float_common.h: In function ‘fast_float::value128 fast_float::full_multiplication(uint64_t, uint64_t)’:
/home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/float_common.h:120:3: error: ‘__uint128_t’ was not declared in this scope
   __uint128_t r = ((__uint128_t)value1) * value2;
   ^~~~~~~~~~~
/home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/float_common.h:120:3: note: suggested alternative: ‘__uint32_t’
   __uint128_t r = ((__uint128_t)value1) * value2;
   ^~~~~~~~~~~
   __uint32_t
/home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/float_common.h:121:25: error: ‘r’ was not declared in this scope
   answer.low = uint64_t(r);
                         ^
/home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/float_common.h:117:63: error: unused parameter ‘value1’ [-Werror=unused-parameter]
 fastfloat_really_inline value128 full_multiplication(uint64_t value1,
                                                               ^~~~~~
/home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/float_common.h:118:63: error: unused parameter ‘value2’ [-Werror=unused-parameter]
                                                      uint64_t value2) {
                                                               ^~~~~~
[  7%] Building CXX object CMakeFiles/c4core.dir/src/c4/format.cpp.o
In file included from /home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/parse_number.h:3:0,
                 from /home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/fast_float.h:44,
                 from /home/runner/work/c4core/c4core/src/c4/charconv.hpp:21,
                 from /home/runner/work/c4core/c4core/src/c4/base64.hpp:9,
                 from /home/runner/work/c4core/c4core/src/c4/base64.cpp:1:
/home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/ascii_number.h: In function ‘fast_float::parsed_number_string fast_float::parse_number_string(const char*, const char*, fast_float::chars_format)’:
/home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/ascii_number.h:102:35: error: useless cast to type ‘int32_t {aka int}’ [-Werror=useless-cast]
       int32_t(p - start_digits - 1); // used later to guard against overflows
                                   ^
/home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/ascii_number.h:150:44: error: useless cast to type ‘int’ [-Werror=useless-cast]
     digit_count -= int(start - start_digits);
                                            ^
/home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/ascii_number.h: In function ‘fast_float::decimal fast_float::parse_decimal(const char*, const char*)’:
/home/runner/work/c4core/c4core/src/c4/ext/fast_float/include/fast_float/ascii_number.h:218:58: error: useless cast to type ‘int32_t {aka int}’ [-Werror=useless-cast]
     answer.decimal_point = int32_t(first_after_period - p);
                                                          ^
cc1plus: all warnings being treated as errors
```

### PR DESCRIPTION
**Pull request by biojppm**
This resolves #33:

- [X] add 32 bit builds to the CI
- [X] compile in VS
- [x]  tests succeed in VS
- [X] compile in gcc
- [x] tests succeed in gcc
- [x]  compile in MinGW
- [x] tests succeed in MinGW
- [x] compile in VS-clang
- [x] tests succeed in VS-clang
- [x] s390x tests succeed
- [x] ppc64le tests succeed

### CODE REVIEW

File: include/fast_float/float_common.h [hunk 1]
```diff
@@ -111,15 +111,31 @@ fastfloat_really_inline value128 full_multiplication(uint64_t value1,
...
+  uint64_t a_hi = a >> 32;
+  uint64_t b_lo = b & lo32;
+  uint64_t b_hi = b >> 32;
+  uint64_t ab_hi =  a_hi * b_hi;
+  uint64_t ab_mid = a_hi * b_lo;
+  uint64_t ba_mid = b_hi * a_lo;
+  uint64_t ab_lo =  a_lo * b_lo;
+  uint64_t carry_bit = ((ab_mid & lo32) + (ba_mid & lo32) + (ab_lo >> 32)) >> 32;
+  answer.high = ab_hi + (ab_mid >> 32) + (ba_mid >> 32) + carry_bit;
+  answer.low = ab_lo + (ab_mid & lo32) + (ba_mid & lo32);
```
Comment (biojppm) [inline include/fast_float/float_common.h:133]: not sure about this. @lemire can you confirm?

File: .github/workflows/vs16-clang-ci.yml [hunk 2]
```diff
@@ -6,6 +6,12 @@ jobs:
   ci:
     name: windows-vs16
     runs-on: windows-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - vs: VS16Win64
+          - vs: VS16Win32
```
Comment (biojppm) [inline .github/workflows/vs16-clang-ci.yml:14]: The `VS16Win32` value was a guess. Apparently [that was correct](https://github.com/lukka/run-cmake/blob/main/action.yml#L40) --- but it turns out [the build is still 64 bit](https://github.com/lemire/fast_float/runs/1403807091?check_suite_focus=true#step:3:21).

Comment (biojppm) [reply .github/workflows/vs16-clang-ci.yml:14]: Would you accept if I changed this to not use the`lukka/run-cmake` action?

File: include/fast_float/float_common.h [hunk 3]
```diff
@@ -62,13 +62,25 @@ struct value128 {
...
+#if defined(_M_X64) || defined(_M_ARM64)
   unsigned long leading_zero = 0;
   // Search the mask data from most significant bit (MSB)
   // to least significant bit (LSB) for a set bit (1).
   if (_BitScanReverse64(&leading_zero, input_num))
     return (int)(63 - leading_zero);
   else
     return 64;
+#else
+    if(!input_num) return 64;
```
Comment (biojppm) [inline include/fast_float/float_common.h:74]: In the original function there is a slight difference in behavior; the VS version defends the call to `_BitScanReverse64()` with an if, suggesting that 0 is a valid input. However, the gcc version uses `__builtin_clzll()` unprotected, so 0 will pass undetected, but the output will be invalid.

I initially went the VS way. As expected, a test asserting that `leading_zeroes(0) == 64` passes in win64 and win32, and [fails in MinGW](https://github.com/lemire/fast_float/runs/1407301137?check_suite_focus=true#step:7:11).

I'm not sure what to do here. Shall I add an if to MinGW? Shall I remove the ifs and the test?

Comment (lemire) [reply include/fast_float/float_common.h:74]: `leading_zeroes(0)` is never called. We do not rely on the fact that `leading_zeroes(0) = 64`. In fact, should 0 be ever be passed to this function, returning 64 would be wrong. There is no way to recover if 0 is passed the function. It cannot and should not happen.

File: .github/workflows/vs16-clang-ci.yml (hunk 2 above)
Comment (lemire) [reply .github/workflows/vs16-clang-ci.yml]: > Would you accept if I changed this to not use thelukka/run-cmake action?

We do not have to use `thelukka`. Maybe, however, we could reach out and report a problem if there is.

File: include/fast_float/float_common.h (hunk 3 above)
Comment (lemire) [reply include/fast_float/float_common.h:74]: We deliberately document it...

```C++
/* result might be undefined when input_num is zero */
```

That is, calling this function with zero is a mistake. 

This should not be checked at runtime (for obvious performance reasons).

Comment (biojppm) [reply include/fast_float/float_common.h:74]: > This should not be checked at runtime (for obvious performance reasons).

agreed.
```cpp
// Should I then turn this...
if (_BitScanReverse64(&leading_zero, input_num))
  return (int)(63 - leading_zero);
else
  return 64;

// ... into this?
return (int)(63 - _BitScanReverse64(&leading_zero, input_num));
```

Comment (lemire) [reply include/fast_float/float_common.h:74]: Looks fine to me.

File: include/fast_float/float_common.h (hunk 1 above)
Comment (lemire) [reply include/fast_float/float_common.h:148]: It looks good at a glance. And it is not obviously inefficient. If the logic is somehow wrong (could happen, there are many variables here), then a test would almost surely reveal the mistake. It is hard to be "almost correct" with these problems. :-)

File: .github/workflows/vs16-clang-ci.yml (hunk 2 above)
Comment (biojppm) [reply .github/workflows/vs16-clang-ci.yml]: https://github.com/lukka/run-cmake/issues/35

Comment (lemire) [reply .github/workflows/vs16-clang-ci.yml]: @biojppm I am sure that the author will appreciate!

**Comment by biojppm** [comment]
@lemire the previous commit changed basictest to use [https://github.com/onqtam/doctest](doctest) for the basic tests. Although I think this is an improvement on the whole, the project is yours and so is the decision whether this. If you do not fully agree with it I'm happy to move back to the previous state.

FWIW, as noted in the commit, I had trouble seeing what was wrong in the CI failures, and that moved me to do this change.

Also, I should add that I'm getting doctest through `FetchContent()` which is available only from cmake 3.11 onwards, and that should be also taken into account.

**Comment by biojppm** [comment]
The good news is that gcc64,v64,vs32 are working ATM. The bad news is that a lot of other things aren't: gcc32 (but that's expected), and vs-clang,mingw (which I didn't look at yet).

For now let's look at gcc32. With detailed output of the test I can [show what's going wrong](https://github.com/lemire/fast_float/runs/1409870045?check_suite_focus=true#step:7:6):

```shell
/home/runner/work/fast_float/fast_float/tests/basictest.cpp:35:
TEST CASE:  full_multiplication

/home/runner/work/fast_float/fast_float/tests/basictest.cpp:28: ERROR: CHECK_EQ( v.low, expected_lo ) is NOT correct!
  values: CHECK_EQ( 2147483648, 9223372036854775808 )
  logged: lhs=0x1 (1)  rhs=0x8000000000000000 (9223372036854775808)
  actualLo=0x80000000 (2147483648)  actualHi=0x0 (0)
  expectedLo=0x8000000000000000 (9223372036854775808)  expectedHi=0x0 (0)
```

So indeed something bad with `full_multiplication()`. In this case `(1<<63)*1`.

**Comment by biojppm** [comment]
The good thing is even though gcc32 fails, the number of failures is actually low: only 13 assertions failed out of 4555, or 0.2%. 

And I think (or I maybe I just want to think) that these failures are related with the faulty `full_multiplication()`, so if we fix that we'll be much closer.

**Comment by lemire** [comment]
> Although I think this is an improvement on the whole

It is fine, but please make sure that it is only enabled if FASTFLOAT_TEST is on. The idea here is that most users will want to use this as a dependency without needing to run the tests, so we don't want to drag along extra dependencies that most people do not need. The tests is just for us (people hacking the code).

I would recommend you sync with our main branch because I made changes so that we can support big-endian encodings. You want to grab that along.

File: tests/CMakeLists.txt [hunk 4]
```diff
@@ -1,3 +1,9 @@
+include(FetchContent)
+
+FetchContent_Declare(doctest
+  GIT_REPOSITORY https://github.com/onqtam/doctest.git
+  GIT_TAG master)
+FetchContent_MakeAvailable(doctest)
```
Comment (lemire) [inline tests/CMakeLists.txt:7]: Please make sure that doctest is not pulled when the tests are not enabled. That is, by default, it should not get pulled.

**Comment by lemire** [comment]
@biojppm Please check this out...
https://github.com/simdjson/simdjson/blob/a9480a768bd80174928fee25ecc33947faad4603/include/simdjson/generic/jsoncharutils.h#L95

It is a working and well tested version of full_multiplication. Yours should work just as well though.

If not, maybe you can gain some karma by correcting stackoverflow. :-)

File: tests/CMakeLists.txt (hunk 4 above)
Comment (biojppm) [reply tests/CMakeLists.txt:7]: That comes free from the main CMakeLists.txt:
```cmake
if(FASTFLOAT_TEST)
  enable_testing()
  add_subdirectory(tests)
endif(FASTFLOAT_TEST)
```

Comment (lemire) [reply tests/CMakeLists.txt:7]: Then we are good. :-)

**Comment by biojppm** [comment]
Almost ready. The only failure ATM is [the MinGW64 pow() function](https://github.com/lemire/fast_float/pull/36/checks?check_run_id=1427847518#step:7:16). I guess this was already failing before this PR?

**Comment by lemire** [comment]
@biojppm That's correct. Not your fault.

**Comment by lemire** [comment]
@biojppm We get the following output under MSYS2:

```
===============================================================================
21
[doctest] test cases:   12 |   11 passed | 1 failed | 0 skipped
22
[doctest] assertions: 2488 | 2487 passed | 1 failed |
23
[doctest] Status: FAILURE!
```

Can we get a verbose output so that we know which test failed? The warning about pow(...-308) is expected, but the test failure is unexpected.

**Comment by lemire** [comment]
@biojppm As part of this PR, would you please add your name to the "AUTHORS" file. It is fine if you don't, but I will then need to do it as part of a separate commit.

I am also giving you direct access to the repo as of now.

(As discussed by email, I am totally open to moving fast_float to its own org so that it is not under my name.)

**Comment by lemire** [comment]
This is effectively done, of course, but I'd like for us to see what the MSYS2 failure is all about and work around it. It is true that MSYS2 has a broken pow function, but that *we* are not broken under MSYS2.

**Comment by biojppm** [comment]
>We get the following output under MSYS2:
>
>===============================================================================
>[doctest] test cases:   12 |   11 passed | 1 failed | 0 skipped
>[doctest] assertions: 2488 | 2487 passed | 1 failed |
>[doctest] Status: FAILURE!
>
> Can we get a verbose output so that we know which test failed? The warning about pow(...-308) is expected, but the test failure is unexpected.

We do: https://github.com/lemire/fast_float/runs/1427997766?check_suite_focus=true#step:7:16 :

```shell
[doctest] doctest version is "2.4.1"
[doctest] run with "--help" for options
===============================================================================
../tests/basictest.cpp:193:
TEST CASE:  powers_of_ten

../tests/basictest.cpp:195: ERROR: CHECK( 1e-308 == std::pow(10, -308) ) is NOT correct!
  values: CHECK( 0.0 == 0.0 )
  logged: On your system, the pow function is busted. Sorry about that.

===============================================================================
[doctest] test cases:   12 |   11 passed | 1 failed | 0 skipped
[doctest] assertions: 2488 | 2487 passed | 1 failed |
[doctest] Status: FAILURE!
```

File: include/fast_float/float_common.h (hunk 1 above)
Comment (biojppm) [reply include/fast_float/float_common.h]: Turns out you already had the working code there :-). Just needed to shift around the feature test macros.

**Comment by biojppm** [comment]
MSYS is now fixed. But I have yet to find a way to get a recent cmake into the travis images.

**Comment by biojppm** [comment]
Phew. This last part was really hard -- I usually find that iterating with travis setup is very time consuming, and in this case it was compounded by the fact that the s390x and ppc64le images are a bit out of the norm.

 But finally it seems this is ready.

Thanks for granting access. Of course, any changes from me will still go through the PR rule. As for the fast_float org, it indeed makes sense if the repo gains traction.

**Comment by lemire** [comment]
I have opened an issue regarding the org.

**Comment by lemire** [comment]
Excellent work. Merging.

**Pull request merged**